docs/config.md
## Configuring Frontik

Here is the list of all options currently supported by Frontik.
All of them can be set in the configuration file (see [frontik_dev.cfg.ex](/frontik_dev.cfg.ex) for example),
but you can pass them to `/usr/bin/frontik` script through command line as well. The exception is `config` option — it makes sense to pass it only through command line.

These options can be set for each Frontik instance (see [options.py](/frontik/options.py)).

| Option name                  | Type    | Default value | Description                                                            |
| ---------------------------- | ------- | ------------  | ---------------------------------------------------------------------- |
| `host`                       | `str`   | `'0.0.0.0'`   | Host value for incoming connections                                    |
| `port`                       | `int`   | `8000`        | Port to listen to                                                      |
| `reuse_port`                 | `bool`  | `True`        | Binds server socket with SO_REUSEPORT option                           |
| `config`                     | `str`   | `None`        | Path to config file                                                    |
| `slow_callback_threshold_ms` | `int`   | `None`        | Enables logging of long-running callbacks                              |
| `app`                        | `str`   | `None`        | Application name  (is used for application identification and as default path to it's class)|
| `app_class`                  | `str`   | `None`        | Path to application class (by default `FrontikApplication` class is used (see [Frontik application structure](/docs/frontik-app.md))) |
| `workers`                    | `int`   | `1`           | Number of worker processes creates using fork. When default value is used, master itself become worker, without fork |
| `init_workers_timeout_sec`   | `int`   | `60`          | Timeout for worker initialization                                      |
| `xheaders  `                 | `bool`  | `False`       | Controls Tornado HTTPServer `xheaders` option                          |
| `tornado_settings`           | `dict`  | `None`        | tornado.web.Application settings                                       |
| `autoreload`                 | `bool`  | `False`       | Restart Frontik after changes in application sources or config files   |
| `debug`                      | `bool`  | `False`       | Enable debug mode                                                      |
| `debug_login`                | `str`   | `None`        | Debug mode login for basic authentication (when `debug=False`)         |
| `debug_password`             | `str`   | `None`        | Debug mode password for basic authentication (when `debug=False`)      |
| `max_active_handlers`        | `int`   | `100`         | Limit the maximum number of simultaneous requests per worker           |
| `datacenter`                 | `str`   | `None`        | Datacenter where current application is running                        |
| `validate_request_id`        | `bool`  | `False`       | Enable check for validity of request_id (32 characters hex-string)     |
| `node_name`                  | `str`   | `False`       | Name of node where frontik application starts                          |
| `common_executor_pool_size`  | `int`   | `10`          | The maximum number of threads that can be used to execute the given calls |
| `stop_timeout`               | `int`   | `3`           | Max time in seconds for gracefully application stopping                |

Logging options:

| Option name                                 | Type    | Default value | Description                                                            |
|---------------------------------------------|---------|---------------|------------------------------------------------------------------------|
| `log_level`                                 | `str`   | `info`        | Python log level                                                       |
| `logformat`                                 | `str`   | see code      | Log entry format for files and syslog                                  |
| `log_dir`                                   | `str`   | `None`        | Log directory location (set to `None` to disable logging to file)      |
| `log_json`                                  | `bool`  | `True`        | Enable JSON logging for files and syslog                               |
| `log_text_format`                           | `str`   | see code      | Log format for files and syslog when JSON logging is disabled          |
| `stderr_log`                                | `bool`  | `False`       | Send log output to stderr (colorized if possible)                      |
| `stderr_format`                             | `str`   | see code      | Log entry format for stderr output                                     |
| `stderr_dateformat`                         | `str`   | see code      | Log entry date format for stderr output                                |
| `syslog`                                    | `bool`  | `False`       | Enables sending logs to syslog                                         |
| `syslog_host`                               | `str`   | `127.0.0.1`   | Syslog host                                                            |
| `syslog_port`                               | `int`   | `None`        | Syslog port. If this value is None, unix socket is used, UDP otherwise |
| `syslog_facility`                           | `str`   | `'user'`      | Syslog facility                                                        |
| `syslog_tag`                                | `str`   | `''`          | Syslog tag                                                             |
| `suppressed_loggers`                        | `list`  | `[]`          | List of logger names to be excluded from debug output                  |
| `sentry_dsn`                                | `str`   | `None`        | Enable Sentry and set Sentry DSN for sending errors                    |
| `statsd_host`                               | `str`   | `None`        | Stats server host for metrics                                          |
| `statsd_port`                               | `int`   | `None`        | Stats server port for metrics                                          |
| `statsd_default_periodic_send_interval_sec` | `int`   | `60`          | Stats default periodic metrics sending interval                        |
| `asyncio_task_threshold_sec`                | `float` | `None`        | Threshold for logging long-running asyncio tasks                       |
| `asyncio_task_critical_threshold_sec`       | `float` | `None`        | Threshold for send to Sentry long-running asyncio tasks                |

HTTP client options:

| Option name                                   | Type    | Default value      | Description                                                                                |
| --------------------------------------------- | ------- | ------------------ | ------------------------------------------------------------------------------------------ |
| `max_http_clients`                            | `int`   | `100`              | Curl max clients option                                                                    |
| `max_http_clients_connects`                   | `int`   | `None`             | Curl max connects option                                                                   |
| `http_client_default_connect_timeout_sec`     | `float` | `0.2`              | Default connect timeout                                                                    |
| `http_client_default_request_timeout_sec`     | `float` | `2.0`              | Default request timeout                                                                    |
| `http_client_default_max_tries`               | `int`   | `2`                | Maximum number of retries per request + 1                                                  |
| `http_client_default_max_timeout_tries`       | `int`   | `1`                | Maximum number of retries due to timeout per request + 1                                   |
| `http_client_default_retry_policy`            | `dict`  | `{599: False, 503: False}` | Conditions when request retry is possible                                                  |
| `http_proxy_host`                             | `str`   | `None`             | HTTP proxy host for Curl HTTP client                                                       |
| `http_proxy_port`                             | `int`   | `3128`             | HTTP proxy port for Curl HTTP client                                                       |
| `http_client_allow_cross_datacenter_requests` | `bool`  | `False`            | Allow requests to different datacenter when no upstream in current datacenter is available |
| `timeout_multiplier`                          | `float` | `1.0`              | Generic timeout multiplier for http requests (useful for testing)                          |

Producers options:

| Option name                         | Type   | Default value | Description                                           |
|-------------------------------------|--------|---------------|-------------------------------------------------------|
| `xsl_root`                          | `str`  | `None`        | Root directory for XSL files                          |
| `xml_root`                          | `str`  | `None`        | Root directory for XML files                          |
| `xsl_cache_limit`                   | `int`  | `None`        | Upper limit for XSL LRU files cache                   |
| `xml_cache_limit`                   | `int`  | `None`        | Upper limit for XML LRU files cache                   |
| `xsl_cache_step`                    | `int`  | `None`        | Increase in weight for XSL cache entry after each get |
| `xml_cache_step`                    | `int`  | `None`        | Increase in weight for XML cache entry after each get |
| `xsl_executor_pool_size`            | `int`  | `1`           | Number of background threads for XSLT processing      |
| `jinja_template_root`               | `str`  | `None`        | Root directory for Jinja templates                    |
| `jinja_template_cache_limit`        | `int`  | `50`          | Upper limit for Jinja templates cache                 |
| `jinja_streaming_render_timeout_ms` | `int`  | `50`          | Upper limit (in msecs) for one iteration of partial Jinja template rendering |

Consul options:

| Option name                                 | Type   | Default value | Description                                                            |
|---------------------------------------------|--------|---------------|------------------------------------------------------------------------|
| `consul_enabled`                            | `bool` | `True`        | Enable Consul features: registration, kv-store, upstreams              |
| `consul_host`                               | `str`  | `127.0.0.1`   | Consul host                                                            |
| `consul_port`                               | `int`  | `None`        | Consul port                                                            |    
| `consul_service_address`                    | `str`  | `None`        | Address of application for Consul registration                         |
| `consul_check_host`                         | `str`  | `None`        | Address for healthcheck application, if not provided, `consul_service_address` is used  |
| `consul_http_check_interval_sec`            | `int`  | `10`          | Interval for making healthcheck request                                |
| `consul_http_check_timeout_sec`             | `int`  | `1`           | Timeout for health check request                                       |
| `consul_tags`                               | `list` | `[]`          | Additional parameters, which will will be sent with Consul registration     |
| `consul_weight_watch_seconds`               | `int`  | `600`         | Max waiting blocking query time for get host weight from Consul        |
| `consul_weight_total_timeout_sec`           | `int`  | `650`         | Timeout for waiting blocking query for getting host weight from Consul |
| `consul_cache_initial_warmup_timeout_sec`   | `int`  | `2`           | Ttl of HTTP session to initialize cache                                |
| `consul_cache_backoff_delay_seconds`        | `int`  | `10`          | Interval to wait until next attempt to fetch info from consul after exception   |
| `consul_consistency_mode`                   | `str`  | `default`     | May set as `default`, `consistent` and `stale`. See Consul documentation   |
| `consul_weight_consistency_mode`            | `str`  | `default`     | Consistency mode for get weigth.                                       |
| `consul_deregister_critical_timeout`        | `str`  | `120h`        | Time, after that service will be de-registered automatically from Consul |
| `upstreams`                                 | `list` | `[]`          | List of service upstreams - services, where current app will send http requests |
| `fail_start_on_empty_upstream`              | `bool` | `True`        | If `True` app will not start if one or more upstreams don't have host addresses |

Opentelemetry options:

| Option name                         | Type   | Default value | Description                                                            |
|-------------------------------------|--------|---------------|------------------------------------------------------------------------|
| `opentelemetry_collector_url`       | `str`  | `127.0.0.1`   | OpenTelemetry collector url                                            |
| `opentelemetry_sampler_ratio`       | `float`| `0.01`        | Probability (between 0 and 1) that a span will be sampled              |
| `opentelemetry_enabled`             | `bool` | `False`       | Enable OpenTelemetry features                                          |

The only option that is mandatory for running Frontik is `app` option — the name of application package.

There are also certain options, that can be defined during application initialization, see
[Configuring Frontik application](/docs/config-app.md) for more details.


debian/changelog
frontik (5.4.4) unstable; urgency=medium

  * HH-85510 remove coding=utf-8
  * HH-85510 remove u prefix
  * HH-85510 fix empty xml in debug

 -- Automated Build <an.sumin@hh.ru>  Wed, 21 Nov 2018 17:24:00 +0300

frontik (5.4.3) unstable; urgency=medium

  * HH-85064 install frontik launcher from setup.py
  * HH-85064 capture warnings by logging module
  * HH-85064 remove asynchronous decorator usages

 -- Automated Build <an.sumin@hh.ru>  Thu, 15 Nov 2018 16:55:00 +0300

frontik (5.4.2) unstable; urgency=medium

  * HH-84940 fix 405 errors

 -- Automated Build <an.sumin@hh.ru>  Tue, 06 Nov 2018 15:53:09 +0300

frontik (5.4.1) unstable; urgency=medium

  * HH-84906 small logging improvements

 -- Automated Build <an.sumin@hh.ru>  Fri, 02 Nov 2018 18:03:53 +0300

frontik (5.4.0) unstable; urgency=medium

  * HH-78387 remove MicroHandler

 -- Automated Build <an.sumin@hh.ru>  Fri, 02 Nov 2018 14:20:13 +0300

frontik (5.3.0) unstable; urgency=medium

  * HH-84844 remove custom HTTPError

 -- Automated Build <an.sumin@hh.ru>  Fri, 02 Nov 2018 14:12:20 +0300

frontik (5.2.0) unstable; urgency=medium

  * HH-84865 add functools.wraps for some wrappers

 -- Automated Build <an.sumin@hh.ru>  Thu, 01 Nov 2018 18:18:33 +0300

frontik (5.1.1) unstable; urgency=medium

  * HH-84820 add slow callbacks monitoring

 -- Automated Build <an.sumin@hh.ru>  Thu, 01 Nov 2018 12:33:15 +0300

frontik (5.1.0) unstable; urgency=medium

  * HH-84737 remove httpcodes module

 -- Automated Build <an.sumin@hh.ru>  Wed, 31 Oct 2018 13:32:52 +0300

frontik (5.0.2) unstable; urgency=medium

  * HH-84709 restore log_blocked_ioloop_timeout setting

 -- Automated Build <an.sumin@hh.ru>  Wed, 31 Oct 2018 13:26:18 +0300

frontik (5.0.1) unstable; urgency=medium

  * HH-84741 remove ioloop blocks logging

 -- Automated Build <an.sumin@hh.ru>  Tue, 30 Oct 2018 16:04:09 +0300

frontik (5.0.0) unstable; urgency=medium

  * HH-84687 use tornado 5

 -- Automated Build <an.sumin@hh.ru>  Mon, 29 Oct 2018 16:52:23 +0300

frontik (4.1.0) unstable; urgency=medium

  * HH-84656 remove simplejson

 -- Automated Build <an.sumin@hh.ru>  Fri, 26 Oct 2018 13:59:28 +0300

frontik (4.0.0) unstable; urgency=medium

  * HH-84575 remove Python 2 from frontik

 -- Automated Build <an.sumin@hh.ru>  Thu, 25 Oct 2018 14:06:02 +0300

frontik (3.4.1) unstable; urgency=medium

  * HH-84347 string json response
  * HH-84091 *args, **kwargs in add_timeout / add_callback
  * HH-84091 early postprocessors -> postprocessors
  * HH-84091 do not output html response in iframe in debug
  * HH-84075 try to fix raven memory leak
  * HH-83696 BREAKING rename frontik.async -> frontik.futures
  * HH-83491 BREAKING return XML node from xml_from_file (old behaviour:
    return list of [comment, node])
  * HH-83491 prepare for future raven releases
  * HH-83491 BREAKING decode text/plain to unicode
  * HH-82583 Convern body to bytes
  * HH-80991 use lower timeouts for preprocessors group tests
  * HH-80991 add tests for preprocessors group
  * HH-80991 introduce asynchGroup future
  * HH-80991 allow to wait preprocessors requests completion
  * HH-79867 Add datacenter and rack handling, fix sync callback
    execution
  * HH-76660 add .hh-release file
  * HH-78634 use pytest, up pycodestyle
  * HH-78543 Add new server to balancing with max load
  * HH-78501 log handler name
  * HH-78415 python3 related fix
  * HH-78120 small compat feature and statsd client python3
    compatibility fix
  * HH-77551 extend frontik.compat
  * HH-77357 add pycurl version to /version
  * HH-77357 update osx travis image
  * HH-77357 move producer settings to frontik.cfg
  * HH-77357 use tornado.concurrent.Future instead of TracebackFuture
  * HH-77483 update install dependencies
  * HH-77483 sync send_error implementation with Tornado 4.5
  * HH-77483 up tornado version
  * HH-77221 migrate to tornado 4.5
  * HH-76515 add max_http_clients_connects option
  * HH-75529 Add gauge statsd metric
  * HH-74480 Fix server weight chenge, and basic documentation
  * HH-73932 Raise custom HTTPError for debug mode
  * HH-74483 remove rendering batch variable
  * HH-74483 extract method
  * HH-74483 allow to set rendering timeout
  * HH-74457 make pretty exceptions handling for stream-templating
  * HH-74434 allow partial jinja templating
  * HH-74353 remove old preprocessors
  * HH-74009 Fix upstream update logging, pass http parameters to
    protobuf client
  * HH-73921 Fix retries logging
  * HH-73794 Change timing metric labels
  * HH-73710 parse data-source-info
  * HH-73662 Fix connect timeout retry, add logging
  * HH-73592 Add non idempotent retry
  * HH-73599 Send service name to statsd with tag rather then with
    prefix
  * HH-73397 Calculate label color for xml
  * HH-73326 Add custom slow start
  * HH-73297 Implement new slow start strategies
  * HH-72947 Fix slow start
  * HH-73201 do not encode json responses recursively
  * HH-73177 Add requests slow start
  * HH-73119 small xml producer optimization
  * HH-72978 Add server randomization, add check for already finished
    page
  * HH-72752 Add connect and request timeout per upstream
  * HH-72655 Add protobuf client
  * HH-72707 Fix requests stats
  * HH-72631 refactor finish_with_postprocessors
  * HH-72649 Add logging params
  * HH-72340 Add http balancing
  * HH-72476 prepare tornado4
  * HH-72475 fix preprocessors early exit
  * HH-72424 fix future return in new preprocessors
  * HH-66084 add preprocessor decorator
  * HH-72025 clean things up after removing Tornado patch
  * HH-70485 remove logging patch
  * HH-71548 fix tornado dependency
  * HH-71548 build python3 deb package
  * HH-70830 update js-libs
  * HH-69223 insane debug refactoring
  * HH-66959 fix debug output if some data not exist
  * HH-68956 disable breadcrumbs in sentry client
  * HH-68476 use global ContextFilter and request-local log handler
  * HH-68298 routing grooming
  * HH-67413 use request_id and handler_name from RequestContext

frontik/options.py
from dataclasses import dataclass, field, fields
import logging.handlers

LOG_DIR_OPTION_NAME = 'log_dir'
STDERR_LOG_OPTION_NAME = 'stderr_log'


@dataclass
class Options:
    app: str = None
    app_class: str = None
    workers: int = 1
    init_workers_timeout_sec: int = 60
    tornado_settings: dict = None
    max_active_handlers: int = 100
    reuse_port: bool = True
    xheaders: bool = False
    validate_request_id: bool = False

    config: str = None
    host: str = '0.0.0.0'
    port: int = 8080
    node_name: str = ''
    common_executor_pool_size: int = 10

    autoreload: bool = False
    stop_timeout: int = 3
    asyncio_task_threshold_sec: float = None
    asyncio_task_critical_threshold_sec: float = None

    log_dir: str = None
    log_level: str = 'info'
    update_log_level_interval_in_seconds: int = 300
    log_json: bool = True
    log_text_format: str = '[%(process)s] %(asctime)s %(levelname)s %(name)s: %(message)s'

    stderr_log: bool = False
    stderr_format: str = '%(color)s[%(levelname)1.1s %(asctime)s %(name)s ' \
                         '%(module)s:%(lineno)d]%(end_color)s %(message)s'
    stderr_dateformat: str = '%H:%M:%S'

    syslog: bool = False
    syslog_host: str = '127.0.0.1'
    syslog_port: int = logging.handlers.SYSLOG_UDP_PORT
    syslog_tag: str = ''
    syslog_facility: str = 'user'

    suppressed_loggers: list = field(default_factory=lambda: ['tornado.curl_httpclient'])

    debug: bool = False
    debug_login: str = None
    debug_password: str = None

    http_client_metrics_kafka_cluster: str = None

    kafka_clusters: dict = field(default_factory=lambda: {})

    statsd_host: str = None
    statsd_port: int = None
    statsd_default_periodic_send_interval_sec: int = 60
    gc_metrics_send_interval_ms: int = None
    long_gc_log_enabled: bool = True
    long_gc_log_threshold_sec: float = 0.01

    xml_root: str = None
    xml_cache_limit: int = None
    xml_cache_step: int = None
    xsl_root: str = None
    xsl_cache_limit: int = None
    xsl_cache_step: int = None
    xsl_executor_pool_size: int = 1
    jinja_template_root: str = None
    jinja_template_cache_limit: int = 50
    jinja_streaming_render_timeout_ms: int = 50

    sentry_dsn: str = None
    sentry_max_breadcrumbs: int = 0

    max_http_clients: int = 100
    max_http_clients_connects: int = None
    send_timeout_stats_interval_ms: int = 60000

    # consul options
    consul_enabled: bool = True
    consul_host: str = '127.0.0.1'
    consul_port: int = None
    consul_service_address: str = None
    consul_check_host: str = None
    consul_http_check_interval_sec: int = 10
    consul_http_check_timeout_sec: float = 1
    consul_tags: list = field(default_factory=lambda: [])
    consul_weight_watch_seconds: int = 600
    consul_weight_total_timeout_sec: int = 650
    consul_cache_initial_warmup_timeout_sec: int = 2
    consul_cache_backoff_delay_seconds: int = 10
    consul_consistency_mode: str = 'default'
    consul_weight_consistency_mode: str = 'default'  # options.consul_consistency_mode
    consul_deregister_critical_timeout: str = '120h'

    # upstream options
    upstreams: list = field(default_factory=lambda: [])
    fail_start_on_empty_upstream: bool = True

    # opentelemetry options
    opentelemetry_collector_url: str = 'http://127.0.0.1:2360'
    opentelemetry_sampler_ratio: float = 0.01
    opentelemetry_enabled: bool = False


options = Options()


def parse_config_file(path):
    config = {}
    with open(path, 'rb') as config_file:
        exec(config_file.read(), config, config)

    for attr in fields(Options):
        new_value = config.get(attr.name, getattr(options, attr.name))
        if attr.type == int and isinstance(new_value, str):
            new_value = int(new_value)

        setattr(options, attr.name, new_value)
